### ISSUE TITLE
WhiteSignal not suitable for integration with different dt

### ISSUE THREAD

**Issue by arvoelke** [nengo/nengo#997]
```
for dt in (0.0001, 0.002):
    with nengo.Network() as model:
        stim = nengo.Node(output=nengo.processes.WhiteSignal(1.0, rms=0.5))
        p = nengo.Probe(stim, synapse=0.02)

    sim = nengo.Simulator(model, dt=dt, seed=0)
    sim.run(1.0)
    
    pylab.figure()
    pylab.title("dt=%s (RMS=%.3f)" % (dt, nengo.utils.numpy.rms(sim.data[p])))
    pylab.plot(sim.trange(), sim.data[p])
    pylab.xlabel("Time (s)")
    pylab.show()
```

![download 1](https://cloud.githubusercontent.com/assets/5420057/14063778/998e8172-f3ae-11e5-97d7-3d711164467d.png)
![download 2](https://cloud.githubusercontent.com/assets/5420057/14063779/9ab1b45c-f3ae-11e5-9408-1f69047b09ff.png)

Even though the RMS of the unfiltered signal is invariant to `dt` (note: it will be `0.5` for both cases when the synapse is `None`), this becomes no longer true once the signal is filtered/integrated.

As mundane as this seems, this is actually a pretty serious issue, because it means that any network using a `WhiteSignal` process as input to either a LIF neuron or a synapse will almost certainly break if the `dt` is decreased significantly. To see why, consider that a LIF neuron with default `tau_rc = 0.02` will lowpass filter its input in basically the same way as a synapse with the same time constant. And then the above example shows that such a filter will change the dynamic range (and so you must rescale the input, or change the RMS).

It took me way too long to debug this issue on my network as I was varying `dt`, because I thought that things would be okay as long as the unfiltered RMS was staying the same (I suspect others would fall into this same trap). We could possibly make the RMS dependent on `dt`, as we do for the `WhiteNoise` process, and then make this the default behaviour since all input signals to a network inevitably end up being filtered somewhere? However something that is not obvious to me is whether this can be made to play nicely with the `high` parameter.

**Comment by hunse** [comment]
So the "problem" is that you're using WhiteSignal as a white signal, when it's really intended to be a band-limited (colored) signal. For what you're doing, you want WhiteNoise. Maybe we should force users to provide a cutoff for WhiteSignal, and furthermore check that that cutoff is below the Nyquist frequency for the given dt. I'm also unhappy with the name WhiteSignal, but we had that discussion a while ago and I think we're stuck with it now.

**Comment by arvoelke** [comment]
My argument is different. It is that in basically every case that we are actually interested in, the signal will end up filtered as in the above example. A quick search for `WhiteSignal` through the codebase shows that this is the case in basically _every instance_ that it is currently used:
 - examples/usage/delay_node.ipynb (synapse filter)
 - examples/learning/learn_product.ipynb (synapse filter)
 - examples/advanced/nef_summary.ipynb (synapse filter)
 - examples/learning/learn_communication_channel.ipynb (synapse filter) 
 - nengo/tests/test_learning_rules.py (synapse filter)
 - nengo/tests/test_neurons.py (synapse filter)
 - nengo/utils/tests/test_neurons.py (LIF filter)
 - nengo/tests/test_synapses.py (LIF filter)

All of these examples will break in principle if you drop `dt` down by a large factor (while the rest of the network gets better).

I'm arguing that if anyone were to try to learn from these examples with a different `dt`, they would be surprised to learn they must rescale the input, since that is not the case for our other processes or for other forms of input.

**Comment by arvoelke** [comment]
That said, I do think your suggestion of forcing a cutoff would indeed fix this. I guess I'm just trying to argue that this is indeed a "real" problem.

**Comment by hunse** [comment]
Yeah, a cutoff fixes it. I was just working on an example to convince you, but if you're convinced then I won't bother. In all the examples I can find our tests/examples, we have a cutoff, so I think they should be fine. The issue only happens when you make the signal actually "white". In WhiteNoise, I have the proper scaling built in so that the signal integrates/filters the same no matter the dt (see the `scale` parameter). This should work in your example above.

Basically the two ways to deal with white noise are to a) band-limit the white noise so it's not actually white, or b) use a proper sqrt(dt) scaling on the white noise when integrating. IIRC, these correspond to the Stratonovich and Ito interpretations of stochastic differential equations, respectively. We have both these options available in Nengo, I think we just need to make it clear to users that they're there and how to use them.

**Comment by arvoelke** [comment]
I also wasn't thinking about the cutoff when I wrote that message, so yes you are totally right. This isn't actually a problem in any of our current examples because they all have sufficiently low cutoffs! My bad.

At this point I'd like to either force the cutoff like you suggested before so that people don't fall into the same trap that I did (I find this tricky to think about...), or clearly document that you almost surely don't want to connect this process to anything if you don't have a cutoff unless you have a solid understanding of this stuff (part of documentation as you're suggesting above). Currently I'm leaning towards if the first if possible.

**Comment by hunse** [comment]
I think it's especially bad that our default right now for WhiteSignal is no cutoff, suggesting that that is a common use case. But there's also no sensible default cutoff I can think of. So I agree, my choice would be to make the cutoff a required parameter so users always have to specify it.

**Comment by arvoelke** [comment]
In terms of sensible default cutoffs, the default refractory period for LIF (2 ms) imposes the constraint of `dt <= 0.002`, which in turn limits the frequency to 500 Hz (or half that?). Such a cutoff would at least ensure that all `dt <= 0.002` produce the same behaviour (things would only break if `dt` is increased past this point with a different neuron model or refractory period). However, this seems a little roundabout, and arbitrarily couples two somewhat distinct constants. So I'm just spitballing here.

**Comment by hunse** [comment]
So my plan wasn't to have the maximum cutoff based on an absolute number. It would depend on dt, specifically `max_cutoff = 0.5 / dt`. So if the user was using a cutoff of 400 Hz with `dt = 0.001`, they'd be fine, but if they bump `dt` up to 0.002 they'd get a build-time error. It would be nicer to have the error happen at validation time but that's not possible if we want it to depend on `dt`.

**Comment by arvoelke** [comment]
So just clarifying: in my last post I was talking about a default cutoff (if we were to go down that road), not a max cutoff. I think the default cutoff would be fixed, while the max cutoff would be `0.5 / dt` as you said. A default of 250 Hz would probably cover most cases, but it's probably still best to have no default.

**Comment by tbekolay** [comment]
Resolved in #999.
